### PR TITLE
Release v0.2.12: add Diff-Aid H3 forecast compatibility

### PR DESCRIPTION
**Pull request by xmarre**
## Summary

Releases Spectrum MiniMax H3 v0.2.12 with the consumer half of a versioned MiniMax H3 external activation-patch compatibility contract, coordinated with **ComfyUI-DiffAid-Patches v1.0.6**.

Spectrum now recognizes deterministic Diff-Aid H3 text-activation modulation without treating its modulation strength as LoRA/model-parameter perturbation, protects hard sigma-window regime transitions with a current-step actual evaluation when required, keeps ER-SDE/offline replay ownership transactional, and fails safe on malformed or inconsistent declared metadata.

Companion producer PR **xmarre/ComfyUI-DiffAid-Patches#9 has been merged and published as Diff-Aid v1.0.6**.

## Runtime contract

The v1 contract is pure scalar/configuration metadata. It identifies the provider, activation-patch kind, MiniMax H3 architecture, unique instance, resolved block indices/model block count, strength, sigma window/ramp, token weighting, conditional-only behavior and modulation scope. Per-call runtime state carries the exact normalized sigma already derived by Diff-Aid's H3 timestep wrapper.

The consumer validates provider/instance identity, schema, architecture/kind/scope, block mappings, scalar ranges, runtime ordering and finite normalized sigma values without mutating producer-owned metadata.

For a partial hard window (`sigma_ramp=0`), an off/on regime change promotes the current step only if it was scheduled as a forecast. The transition is committed only after successful finalization, so retry/abort/rollback cannot advance it early. Multiple contracts crossing together still require at most one actual evaluation. Full `[0,1]` windows have no interior transition; smooth ramps remain continuous.

## Model-aware risk correction

Real testing found and fixed a development regression in the first integration: a five-block Diff-Aid patch at strength `0.5` produced raw structural telemetry `2.112114`, which was incorrectly folded into Spectrum's parameter-space model-aware patch prior. That saturated `patch_risk=1.0`, changed the normal 20-step run to 16 actual + 4 forecast steps, and added 10 unnecessary model-aware NFEs.

The finalized semantics keep this magnitude as separate external-patch telemetry and cache identity, while the calibrated parameter-space `patch_perturbation`, final-block perturbation, profile confidence and forecast-risk prior remain based on the underlying model/parameter patches. Normal online trajectory evidence still participates in model-aware scheduling.

## Coordinated real MiniMax H3 validation

Validated in restarted ComfyUI on native MiniMax H3, ER-SDE, 20 steps, `model_aware_mode=full`, using Diff-Aid blocks `1,13,25,37,50` at strength `0.5`:

| Run | Diff-Aid window | Actual / forecast | Sampler time | Result |
|---|---|---:|---:|---|
| Spectrum control | inactive | 11 / 9 | 187.236 s | 0 model-aware extra NFEs |
| Diff-Aid full hard window | `[0.0,1.0]` | 11 / 9 | 189.001 s | 0 transitions, 0 forced actuals, 0 extra NFEs |
| Diff-Aid partial hard window | `[0.0,0.95]` | 11 / 9 | 184.946 s | transition at `0.947368` on step 8, already actual, 0 extra NFEs |

All three runs preserve the normal Spectrum 11-actual / 9-forecast NFE budget. The compatibility path no longer introduces a meaningful sampling-speed penalty.

The matched multi-shot media test also retained Diff-Aid's stronger prompt adherence. Full-window activation changed the intended shot transition into a more continuous shot, while `sigma_end=0.95` restored the intended cut between the first and later shots while keeping the prompt-enhancement effect. This is documented as an empirical workflow result, not a universal default.

## Safety / maintainability

- malformed declared metadata fails safe to all-actual sampling for the run instead of aborting a valid generation;
- external runtime validation is single-source and non-mutating;
- runtime activation contracts do not masquerade as parameter patch keys;
- compat/hardening install order is enforced;
- teardown preserves the underlying post-run safety hook as a distinct callable;
- offline replay does not re-promote already handled hard transitions;
- a transaction-parity tripwire locks primary, retry, ER-SDE-consumption, finalize and abort ordering to the native Spectrum transaction;
- raw external structural magnitude remains available in debug/run-summary telemetry without polluting calibrated model-aware patch risk.

## Release metadata

- Spectrum package version: **0.2.12**
- required producer for this compatibility path: **ComfyUI-DiffAid-Patches 1.0.6+**
- README updated for the released workflow and validated behavior
- release notes updated for v0.2.12
- source-checkout replay/generic-correction calibration provenance fallback versions advanced to `0.2.12`

The branch has been reconciled with the current `main` after the post-v0.2.11 documentation-only wording updates.

## Validation

Final release-head CI is green across Spectrum's four-revision pinned ComfyUI matrix, including wheel build, native-H3 forecaster smoke test, Ruff, `compileall`, focused external compatibility tests, transaction-order parity coverage and the native H3 suite. CodeRabbit is green and all inline review findings are resolved.

### CODE REVIEW

File: comfyui_spectrum_h3/__init__.py [hunk 1]
```diff
@@ -30,6 +32,15 @@
...
 install_er_sde_tail_policy()
+install_external_patch_compat()
+install_external_patch_hardening()
+
+# External compatibility extends the already-installed safe end-run wrapper.
+# Keep the postrun module's installed-function identity synchronized so its
+# teardown invariant and tests continue to describe the effective runtime hook.
+from . import postrun_safety as _postrun_safety
+
+_postrun_safety._safe_end_run = SpectrumH3Runtime.end_run
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/__init__.py]: @coderabbitai Valid. I removed the `_safe_end_run` rebinding, added `postrun_safety.EFFECTIVE_END_RUN_HOOK`, and updated the teardown test to verify both layers: `SpectrumH3Runtime.end_run is EFFECTIVE_END_RUN_HOOK` and `external_patch_compat._ORIGINAL_END_RUN is postrun_safety._safe_end_run`. That keeps the inner safety implementation identifiable while still checking the effective outer compat wrapper.

File: comfyui_spectrum_h3/external_patch_compat.py [hunk 2]
```diff
@@ -0,0 +1,1123 @@
...
+        sigma = _finite_number(
+            value.get("normalized_sigma"),
+            name=f"runtime_state[{descriptor.instance_id}].normalized_sigma",
+        )
+        if not 0.0 <= sigma <= 1.0:
+            raise ExternalPatchContractError(
+                f"runtime normalized sigma for {descriptor.instance_id!r} is outside [0, 1]"
+            )
+        normalized.append(sigma)
+    return tuple(normalized)
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/external_patch_compat.py:539]: @coderabbitai Valid. `compat._runtime_entries` now keeps `(provider, value)` in local tuples and never writes `__provider_checked` into producer-owned metadata. I removed the duplicate validator and rebinding from `external_patch_hardening.py`; the non-mutation regression test now calls `compat._runtime_entries` directly, so the shipped source of truth is the code under test.

File: comfyui_spectrum_h3/external_patch_compat.py [hunk 3]
```diff
@@ -0,0 +1,1123 @@
...
+                step=decision["step_id"],
+                retry=True,
+            )
+            result = consume_er_sde_increment(result, retry_decision)
+            runtime.finalize_step(decision["run_id"], decision["step_id"])
+            return result
+    except BaseException:
+        if runtime.active_step_id == decision["step_id"]:
+            runtime.abort_step(decision["run_id"], decision["step_id"])
+        raise
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/external_patch_compat.py:988]: @coderabbitai Valid maintainability risk. I used the fallback you explicitly allowed rather than refactoring the transaction-critical sampling path in this review pass: `test_predict_noise_wrapper_transaction_parity_tripwire` compares the critical transaction operations in the captured native wrapper and compat wrapper (begin, model-options injection, ER-SDE consume/clear, retry, offline transition logging, finalize, abort) and also enforces that compat's effective-step logging happens after the model call. A future transaction edit now trips the test and forces the two paths to be reconciled. This keeps the patch narrow while avoiding an unrequested rewrite of `sampling.predict_noise_wrapper`.

File: comfyui_spectrum_h3/external_patch_hardening.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        normalized.append(sigma)
+    return tuple(normalized)
+
+
+def _profile_builder_with_runtime_key_semantics(*args, **kwargs):
+    assert _ORIGINAL_EXTERNAL_PROFILE_BUILDER is not None
+    profile = _ORIGINAL_EXTERNAL_PROFILE_BUILDER(*args, **kwargs)
+    # active_patch_keys is the existing parameter-patch key count. Runtime
+    # activation contracts have their own explicit counter and no parameter keys.
+    return replace(profile, active_patch_keys=int(profile.base.active_patch_keys))
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/external_patch_hardening.py]: @coderabbitai Valid. `active_patch_keys` is now set from `base.active_patch_keys` at the compat profile source. The hardening `dataclasses.replace` wrapper and its profile-builder rebinding were deleted. Runtime activation contracts still contribute to `active_patch_count` / `recognized_runtime_patch_count`, without being misreported as parameter patch keys; the compat profile test now asserts this directly.

File: comfyui_spectrum_h3/external_patch_hardening.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    compat._runtime_entries = _runtime_entries_without_mutation
+    compat._external_profile_from_base = _profile_builder_with_runtime_key_semantics
+    model_aware.get_model_forecastability_profile = (
+        get_model_forecastability_profile_fail_safe
+    )
+    sampling.get_model_forecastability_profile = (
+        get_model_forecastability_profile_fail_safe
+    )
+    model_aware.clear_model_profile_cache = _clear_all_model_profile_caches
+    _INSTALLED = True
```
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/external_patch_hardening.py:57]: @coderabbitai Valid. `install_external_patch_hardening()` now fails explicitly unless compat is already installed and `_ORIGINAL_PROFILE_LOOKUP` has been initialized. I also added a regression test for inverted installer order. The source-level validator/profile wrappers removed by the other findings are no longer part of this installer.

File: README.md [hunk 6]
```diff
@@ -206,6 +206,16 @@ That is 11 actual evaluations and 9 forecasts. Fallbacks, model-aware scheduling
...
+
+A full `[0,1]` window has no interior transition and adds no compatibility NFE. Smooth `sigma_ramp>0` modulation remains continuous and does not force refreshes solely because its gain changes. The hard-boundary guard remains active when `model_aware_mode=off` because it is a forecast-correctness rule rather than an optional scheduling heuristic.
+
+This mechanism is independent of the native ER-SDE stochastic-state compensation described above. Diff-Aid is deterministic activation modulation; Spectrum does not reuse ER-SDE stochastic compensation for it. Spectrum also does **not** post-hoc multiply or otherwise scale the forecasted `[audio | video]` target feature to imitate Diff-Aid. The external modulation occurs before selected transformer blocks and is transformed nonlinearly by the remaining network, so the compatible action at a declared discontinuity is to acquire a real anchor.
```
[the remark this answers is not in the document]
Comment (xmarre) [reply README.md:231]: @coderabbitai Valid. The README now states that the producer half is pending in `ComfyUI-DiffAid-Patches` PR #9 and is not present in the current Diff-Aid v1.0.5 release. It also states that released Diff-Aid builds do not expose this compatibility path until that companion contract lands in a release.

File: tests/test_external_patch_compat.py [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+        replay=True,
+    )
+    run = compat._compat_state(fake).run
+    run.committed_active = (True,)
+    fake.set_step(4, "replay")
+    compat.observe_external_patch_runtime(fake, runtime_options(runtime_entry(sigma=0.50)))
+    assert fake._step.mode == "replay"
+    assert run.transitions == 0
+    assert run.forced_actuals == 0
+
```
[the remark this answers is not in the document]
Comment (xmarre) [reply tests/test_external_patch_compat.py:557]: @coderabbitai Valid. The replay regression test now uses the production step mode `forecast` with `replay=True`, and asserts the mode remains `forecast` with zero transition/forced-actual counters. That now tests replay suppression on a real forecast-mode step.

**Comment by xmarre** [comment]
@coderabbitai Re nitpick `tests/test_external_patch_hardening.py:32-49`: addressed the underlying invariant with the stronger single-source fix. The suggested binding-identity assertion is no longer applicable because `_runtime_entries_without_mutation` was removed entirely. The existing regression test now invokes `compat._runtime_entries` directly and verifies the producer-owned metadata is unchanged, so it tests the actual implementation used by `observe_external_patch_runtime` without depending on installer rebinding.

**Comment by xmarre** [comment]
@coderabbitai Re nitpick `external_patch_compat.py:402-405`: addressed. The effective cache key is now built once by `_external_cache_key(...)`; `get_model_forecastability_profile_with_external_patches` uses that value for lookup/storage and passes the same tuple into `_external_profile_from_base`, removing the duplicated key construction.

**Comment by xmarre** [comment]
@coderabbitai Re nitpick `tests/test_external_patch_compat.py:86-88`: addressed. The three sigma-window boundary cases now match the complete contract invariant `r"0 <= sigma_start <= sigma_end <= 1"`, including `sigma_end=1.1`, so the tests assert the actual validation rule rather than a misleading field-name substring.

**Comment by xmarre** [comment]
@coderabbitai Re nitpick `tests/test_external_patch_compat.py:494-502`: addressed. After the initial malformed runtime state forces fail-safe, the test now starts a later forecast step, asserts `observe_external_patch_runtime(...) is False`, and verifies `contract_failures` remains exactly 1, covering the run-wide latch and preventing repeated failure/promotion accounting.

**Comment by xmarre** [comment]
@coderabbitai Re nitpick `.github/workflows/tests.yml:107-121`: addressed. The focused compatibility suite still runs first for a fast signal; the later full `pytest` command now `--ignore`s those same eight files, so they execute once per matrix job rather than twice.

File: comfyui_spectrum_h3/external_patch_compat.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (xmarre) [reply comfyui_spectrum_h3/external_patch_compat.py:988]: @coderabbitai Follow-up addressed. I added `test_external_patch_transaction_parity.py` to enforce the critical operation order, not only token counts. It checks both native and compat wrappers through primary execute → offline transition log → ER-SDE consume → finalize, then retry preparation → retry execute → log → ER-SDE consume → finalize → outer abort handling. It also separately checks the compat-only effective-mode log is between model execution and solver-side consumption on both primary and retry paths. The new parity test is included in the focused compatibility CI suite.

**Comment by xmarre** [comment]
Paired-contract note: the hard-window precision mismatch was fixed on producer PR `xmarre/ComfyUI-DiffAid-Patches#9` at head `77fd71bb6196ede873f85d82e189c138f554f6a7`.

The fix covers both operands Spectrum classifies: Diff-Aid now publishes the reconstructed runtime `normalized_sigma` at float32 precision **and** publishes `sigma_start` / `sigma_end` at their effective float32 values for hard (`sigma_ramp == 0`) windows. This is necessary because rounding only the runtime coordinate would still leave Spectrum comparing it against an unquantized Python descriptor boundary.

No consumer code change is required in #56 for this case: its inclusive Python `active_at()` comparison now receives producer values in the same effective float32 domain that Diff-Aid actually used. Producer regression coverage includes the concrete `0.251` ULP case, and #9 CI is green (`55 passed`). Real paired hard-window testing can use the current heads.
